pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -240,3 +252,91 @@ func (krgm *keyspaceResourceGroupManager) getServiceLimiter() *serviceLimiter {
 	defer krgm.RUnlock()
 	return krgm.sl
 }
+
+func (krgm *keyspaceResourceGroupManager) getOrCreateRUTracker(name string) *ruTracker {
+	rt := krgm.getRUTracker(name)
+	if rt == nil {
+		krgm.Lock()
+		// Double check the RU tracker is not created by other goroutine.
+		rt = krgm.ruTrackers[name]
+		if rt == nil {
+			rt = newRUTracker(defaultRUTrackerTimeConstant)
+			krgm.ruTrackers[name] = rt
+		}
+		krgm.Unlock()
+	}
+	return rt
+}
+
+func (krgm *keyspaceResourceGroupManager) getRUTracker(name string) *ruTracker {
+	krgm.RLock()
+	defer krgm.RUnlock()
+	return krgm.ruTrackers[name]
+}
+
+// ruTracker is used to track the RU consumption within a keyspace.
+// It uses the algorithm of time-aware exponential moving average (EMA) to
+// sample and calculate the real-time RU/s of each resource group. The main
+// reason for choosing this EMA algorithm is that conventional EMA algorithms or
+// moving average algorithms over a time window cannot handle non-fixed frequency
+// data sampling well. Since the reporting interval of RU consumption depends on
+// the RU consumption rate of the workload, it is necessary to introduce a time
+// dimension to calculate real-time RU/s more accurately.
+type ruTracker struct {
+	syncutil.RWMutex
+	// beta = ln(2) / τ, τ is the time constant which can be thought of as the half-life of the EMA.
+	// For example, if τ = 5s, then the decay factor calculated by e^{-β·Δt} will be 0.5 when Δt = 5s,
+	// which means the weight of the "old data" is 0.5 when the elapsed time is 5s.
+	beta           float64
+	lastSampleTime time.Time
+	lastEMA        float64
+}
+
+func newRUTracker(timeConstant time.Duration) *ruTracker {
+	return &ruTracker{
+		beta: math.Log(2) / timeConstant.Seconds(),
+	}
+}
+
+// Sample the RU consumption and calculate the real-time RU/s as `lastEMA`.
+// - `now` is the current time point to sample the RU consumption.
+// - `totalRU` is the total RU consumption within the `dur`.
+// - `dur` is the time cost to run out of the `totalRU`.
+func (rt *ruTracker) sample(now time.Time, totalRU float64, dur time.Duration) {
+	rt.Lock()
+	defer rt.Unlock()
+	// If `dur` is not greater than 0, skip this record.
+	if dur <= 0 {
+		return
+	}
+	// Calculate the average RU/s within the `dur`.
+	ruPerSec := math.Max(0, totalRU) / dur.Seconds()
+	// If the last sample time is not set, set the last EMA directly.
+	if rt.lastSampleTime.IsZero() {
+		rt.lastEMA = ruPerSec
```

Comment:
Do we need to check minSampledRUPerSec here?

Reply:
`ruPerSec` is calculated from actual consumption, so it is unlikely to be an extremely small value. I don't think it's necessary to perform additional checks here.